### DISCUSSION TITLE
Polyline has wrong position after serializing and deserializing

### DISCUSSION THREAD

**Discussion by MethDamon** [fabricjs/fabric.js#7158] [Q&A]
Please see the problem in this JsFiddle: https://jsfiddle.net/xp1sz0je/56/

I am pretty sure that I am doing something wrong, but I can't really see what. If someone could point me in the right direction, that would be great. Thanks!

**Comment by asturur** [comment]
the bounding box of your polyline is wrong:

![image](https://user-images.githubusercontent.com/1194048/120917447-a97c3c80-c6af-11eb-9313-f29e511859a6.png)

That makes serialization / deserialization behave not as expected.

You should run: `line._setPositionDimensions` after adding the points.

https://jsfiddle.net/cypb2L7o/

that would make the polyline position immediately different, but then working good with serialize / deserialize.

**Comment by i-prikot** [comment]
Hi.
@asturur, very good solution. I've tweaked the code.
```
fabric.Polyline.prototype.setPositionDimensions = function() {
    this._setPositionDimensions(this);
}
```
@MethDamon, example for you https://jsfiddle.net/i_prikot/dsz86v30/4/

P.S. If you specify points in the initialization, then no additional code is needed.

**Comment by MethDamon** [comment]
> Hi.
> @asturur, very good solution. I've tweaked the code.
> 
> ```
> fabric.Polyline.prototype.setPositionDimensions = function() {
>     this._setPositionDimensions(this);
> }
> ```
> 
> @MethDamon, example for you https://jsfiddle.net/i_prikot/dsz86v30/4/
> 
> P.S. If you specify points in the initialization, then no additional code is needed.

I need to add points on the go interactively, how can I make sure the polyline stays in the same position?

**Comment by i-prikot** [comment]
> > Hi.
> > @asturur, very good solution. I've tweaked the code.
> > ```
> > fabric.Polyline.prototype.setPositionDimensions = function() {
> >     this._setPositionDimensions(this);
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > @MethDamon, example for you https://jsfiddle.net/i_prikot/dsz86v30/4/
> > P.S. If you specify points in the initialization, then no additional code is needed.
> 
> I need to add points on the go interactively, how can I make sure the polyline stays in the same position?

You should run:` line.setPositionDimensions(); `after adding the points.

**Comment by MethDamon** [comment]
> > > Hi.
> > > @asturur, very good solution. I've tweaked the code.
> > > ```
> > > fabric.Polyline.prototype.setPositionDimensions = function() {
> > >     this._setPositionDimensions(this);
> > > }
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > @MethDamon, example for you https://jsfiddle.net/i_prikot/dsz86v30/4/
> > > P.S. If you specify points in the initialization, then no additional code is needed.
> > 
> > 
> > I need to add points on the go interactively, how can I make sure the polyline stays in the same position?
> 
> You should run:`line.setPositionDimensions();`after adding the points.

I am doing that, but the points do not end up where I want them (where the user clicks). What are the coordinates that I should add, do I have to add the points with their absolute coordinates?

**Comment by asturur** [comment]
Yes positioning is never easy for those things.
We should have an example to create a brush for drawing polygons.
The issue is that points are offset from the center of the shape.
But the center of the shape moves for each point you add.

So once you run `line._setPositionDimensions()` you should move the final object in the right position.

This position can be found in in this way ( is complicated ):
you remember where you would like that the first ( or any ) point of the line is positioned on the canvas. (finalX, finalY).
you take that point and you calculate its relative position in % inside the shape
you use setPositionByOrigin method that takes a position and an origin, passing the origin as the % position.

let me see if i can add to the fiddle in few minutes
https://jsfiddle.net/dw56tvpa/2/

ok so choosing a reference point from the mouse click in absolute coordinates and then forcing the path to realing there after having done _setPostionAndDimensions is the way to go imho.
